dig.go

```diff
@@ -498,6 +523,18 @@ func (c *Container) Invoke(function interface{}, opts ...InvokeOption) error {
 	return nil
 }
 
+func (c *Container) verifyAcyclic() error {
+	visited := make(map[key]struct{})
+	for _, n := range c.nodes {
+		if err := detectCycles(n, c, nil /* path */, visited); err != nil {
+			return errWrapf(err, "this function introduces a cycle")
+		}
+	}
+
+	c.isVerifiedAcyclic = true
```

Comment:
nit: I would set this in the `if` clause above where you're checking it so
that it reads like this:

```
if !isVerified {
  verify()
  isVerified = true
}
```

Reply:
I'm not sure I agree. The `if` clause above is only testing starting from a single node. We don't really know that the container is verifiably acyclic until after all nodes have been tested.

Reply:
Oh, my comment may have been confusing. I was talking about the one in Invoke.

```
if !c.isVerifiedAcyclic {
    if err := c.verifyAcyclic(); err != nil {
        return err
    }
}
```

Right now, we check `isVerifiedAcyclic` in Invoke, then verifyAcylic sets it.
I'm suggesting,

```
if !c.isVerifiedAcyclic {
    if err := c.verifyAcyclic(); err != nil {
        return err
    }
    c.isVerifiedAcyclic = true
}
```